src/Blammo/Logging/Logger.hs

```diff
@@ -49,7 +51,34 @@ import System.Log.FastLogger.Compat
   , newStderrLoggerSetN
   , newStdoutLoggerSetN
   )
-import UnliftIO.Exception (throwString)
+import UnliftIO (MonadUnliftIO)
+import UnliftIO.Exception (finally, throwString)
+
+-- | Initialize logging, pass a 'Logger' to the callback, and clean up at the end.
+--
+-- Applications should avoid calling this more than once in their lifecycle.
+withLogger :: MonadUnliftIO m => LogSettings -> (Logger -> m a) -> m a
+withLogger settings f = do
+  logger <- newLogger settings
+  f logger `finally` flushLogStr logger
+
+-- | Write a message to the 'Logger', unless the logger's filter options
+--   reject it based on its 'LogSource' and 'LogLevel'
+runLogAction
+  :: (MonadIO m, ToLogStr msg)
+  => Logger
+  -> Loc
+  -> LogSource
+  -> LogLevel
+  -> msg
+  -> m ()
+runLogAction logger loc source level msg =
+  liftIO $
+    when (lShouldLog logger source level) $
+      defaultOutputWith options loc source level (toLogStr msg)
+ where
+  options = defaultOutputOptions $ \logLevel bytes ->
+    pushLogStrLn logger $ toLogStr $ getLoggerReformat logger logLevel bytes
```

Comment:
Might able to nip off the `logLevel` and `bytes` arguments here too, but I can't do it in my head.

Reply:
I really wish the log function weren't curried. If it were just a record with (loc,src,level,message) then this stuff would be easier to golf.